pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,396 @@
+	newNLM := newObj.(*v1alpha2.NodeLatencyMonitor)
+	klog.InfoS("NodeLatencyMonitor updated", "NodeLatencyMonitor", klog.KObj(newNLM))
+
+	if oldNLM.GetGeneration() == newNLM.GetGeneration() {
+		return
+	}
+
+	if err := m.updateLatencyConfig(newNLM); err != nil {
+		klog.ErrorS(err, "Failed to update latency config")
+	}
+}
+
+func (m *MonitorTool) updateLatencyConfig(nlm *v1alpha2.NodeLatencyMonitor) error {
+	// Parse the ping interval and timeout
+	pingInterval, err := time.ParseDuration(nlm.Spec.PingInterval)
+	if err != nil {
+		return err
+	}
+	pingTimeout, err := time.ParseDuration(nlm.Spec.PingTimeout)
+	if err != nil {
+		return err
+	}
+
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+		Timeout:  pingTimeout,
+		Limit:    nlm.Spec.PingConcurrentLimit,
+	}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+
+	return nil
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *MonitorTool) onNodeLatencyMonitorDelete(obj interface{}) {
+	klog.InfoS("NodeLatencyMonitor deleted", "NodeLatencyMonitor")
+
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{Enable: false}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+}
+
+func getRate(count int, d time.Duration, burst int) *rate.Limiter {
+	r := float64(count) / (d.Seconds() * 0.9)
+	return rate.NewLimiter(rate.Limit(r), burst)
+}
+
+func (m *MonitorTool) pingAll() {
+	// Get all node internal/external IP.
+	nodeIPs := m.latencyStore.ListNodeIPs()
+
+	// A simple rate limiter
+	limiter := getRate(len(nodeIPs), m.latencyConfig.Interval, m.latencyConfig.Limit)
+	limitGroup := make(chan bool, int(m.latencyConfig.Limit))
+
+	klog.InfoS("Start to ping all nodes")
+	wg := sync.WaitGroup{}
+	for name, toIPs := range nodeIPs {
+		for _, toIP := range toIPs {
+			limitGroup <- true
+			limiter.Wait(context.Background())
+			wg.Add(1)
+
+			go func(toIP net.IP, name string) {
+				ctx, cancel := context.WithTimeout(context.Background(), m.latencyConfig.Timeout)
+				defer cancel()
+				defer wg.Done()
+
+				ok, rtt := m.pingNode(ctx, toIP)
+				if ok {
+					m.latencyStore.UpdateConnByKey(name, &Connection{
+						ToIP:        toIP,
+						Latency:     rtt,
+						Status:      true,
+						LastUpdated: time.Now(),
+					})
+				} else {
+					m.latencyStore.UpdateConnByKey(name, &Connection{
+						ToIP:        toIP,
+						Latency:     0,
+						Status:      false,
+						LastUpdated: time.Now(),
+					})
+				}
+
+				<-limitGroup
+
+			}(toIP, name)
+		}
+	}
+
+	wg.Wait()
+
+	// TODO: Print all connection status for debug
+	klog.InfoS("Finish to ping all nodes")
+	conns := m.latencyStore.ListConns()
+	for _, conn := range conns {
+		klog.InfoS("Connection status", "Connection", conn)
+	}
+}
+
+func (m *MonitorTool) pingNode(ctx context.Context, addr net.IP) (bool, time.Duration) {
```

Comment:
I maintain what I said before: I think it would be better to have a single socket that is re-used. I'm attaching a test program that exemplifies what I mean:
```go
package main

import (
	"fmt"
	"net"
	"os"
	"os/signal"
	"syscall"

	"golang.org/x/net/icmp"
	"golang.org/x/net/ipv4"
)

var icmpSeq uint16 = 0

func getICMPSeq() uint16 {
	icmpSeq++
	return icmpSeq
}

func sendPing(socket net.PacketConn, addr net.IP) error {
	ip := &net.IPAddr{IP: addr}
	body := &icmp.Echo{
		ID:   os.Getpid() & 0xffff, // Use the current process ID as the ICMP ID
		Seq:  int(getICMPSeq()),
		Data: []byte("hello"),
	}
	msg := icmp.Message{
		Type: ipv4.ICMPTypeEcho,
		Code: 0,
		Body: body,
	}

	msgBytes, err := msg.Marshal(nil)
	if err != nil {
		return err
	}
	if _, err = socket.WriteTo(msgBytes, ip); err != nil {
		return err
	}
	fmt.Println("Sent echo request")
	return nil
}

func run(stopCh chan bool) {
	const IPv4ProtocolICMPRaw = "ip4:icmp"
	socket, err := icmp.ListenPacket(IPv4ProtocolICMPRaw, "0.0.0.0")
	if err != nil {
		fmt.Printf("Failed to create ICMP socket: %v\n", err)
		os.Exit(1)
	}

	go func() {
		readBuffer := make([]byte, 1024)
		for {
			_, peer, err := socket.ReadFrom(readBuffer)
			if err != nil {
				fmt.Printf("Failed to read ICMP message: %v\n", err)
				continue
			}
			msg, err := icmp.ParseMessage(1, readBuffer)
			if err != nil {
				fmt.Printf("Invalid message: %v\n", err)
				continue
			}
			fmt.Printf("Received reply from %s: %v\n", peer, msg)
		}
	}()

	sendPing(socket, net.IPv4(8, 8, 8, 8))
	sendPing(socket, net.IPv4(8, 8, 4, 4))
	<-stopCh
}

func main() {
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	stopCh := make(chan bool, 1)
	go func() {
		<-sigs
		close(stopCh)
	}()
	run(stopCh)
	<-stopCh
}
```

The RTT can be calculated by embedding the "start time" in the echo request body.

With an implementation like this, I would also just remove the `PingTimeout` and `PingConcurrentLimit` parameters from the API. They don't mean much with this change (we can revisit the concurrency later), and they make the PR more complex. cc @Dyanngg 

For each peer, we then store "send time of last request", "receive time of last reply", "last measured rtt"

Reply:
Ok, so we need to split `request` and `response`, and mark different request with unique id, and update response time to `NodeIPLatencyEntry`.

I will update my code later.

Reply:
Current test log, only two socket(IPv4 and IPv6) send echo request, and receive message then update to nodeIPLatencyMap.

```bash
I0417 16:57:12.511166       1 monitor.go:255] "ICMP data" Data="2024-04-17T16:57:12.510670246Z\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
I0417 16:57:12.511177       1 monitor.go:262] "Received ICMP echo message" Echo={"ID":1,"Seq":5,"Data":"MjAyNC0wNC0xN1QxNjo1NzoxMi41MTA2NzAyNDZaAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="}
I0417 16:57:12.511192       1 monitor.go:263] "Received ICMP echo message" Entry={"SeqID":6,"LastSendTime":"2024-04-17T16:57:11.733733007Z","LastRecvTime":"2024-04-17T16:57:11.73565492Z","LastMeasuredRTT":1921912}
E0417 16:57:12.511197       1 monitor.go:267] "Failed to match seqID"
I0417 16:57:21.733827       1 monitor.go:337] "Finish to ping all nodes"
I0417 16:57:21.733916       1 monitor.go:318] "Start to ping all nodes"
I0417 16:57:21.733950       1 monitor.go:340] "NodeIPLatency status" Key="10.244.1.1" Entry={"SeqID":6,"LastSendTime":"2024-04-17T16:57:11.733733007Z","LastRecvTime":"2024-04-17T16:57:11.73565492Z","LastMeasuredRTT":1921912}
I0417 16:57:21.733972       1 monitor.go:340] "NodeIPLatency status" Key="10.244.0.1" Entry={"SeqID":4,"LastSendTime":"2024-04-17T16:57:11.733593378Z","LastRecvTime":"2024-04-17T16:57:11.733975859Z","LastMeasuredRTT":382467}
I0417 16:57:21.733985       1 monitor.go:340] "NodeIPLatency status" Key="10.244.2.1" Entry={"SeqID":5,"LastSendTime":"2024-04-17T16:57:11.733664621Z","LastRecvTime":"2024-04-17T16:57:11.735432833Z","LastMeasuredRTT":1768222}
```